gce_workload_certs_refresh/main.go

```diff
@@ -0,0 +1,290 @@
+type WorkloadTrustedRootCerts struct {
+	Status           string
+	RootCertificates map[string]RootCertificate
+}
+
+// UnmarshalJSON is a custom JSON unmarshaller for WorkloadTrustedRootCerts
+func (wtrc *WorkloadTrustedRootCerts) UnmarshalJSON(b []byte) error {
+	tmp := map[string]json.RawMessage{}
+	err := json.Unmarshal(b, &tmp)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(tmp["status"], &wtrc.Status); err != nil {
+		return err
+	}
+
+	wtrc.RootCertificates = map[string]RootCertificate{}
+	rcs := map[string]json.RawMessage{}
+	if err := json.Unmarshal(tmp["rootCertificates"], &rcs); err != nil {
+		return err
+	}
+
+	for domain, value := range rcs {
+		rc := RootCertificate{}
+		err := json.Unmarshal(value, &rc)
+		if err != nil {
+			return err
+		}
+		wtrc.RootCertificates[domain] = rc
+	}
+
+	return nil
+}
+
+// RootCertificate represents a Root Certificate in metadata
+type RootCertificate struct {
+	Metadata            Metadata
+	RootCertificatesPem string
+}
+
+// Metadata represents Metadata in metadata
+type Metadata struct {
+	WorkloadCredsDirPath string
+}
+
+func main() {
+	ctx := context.Background()
+
+	opts := logger.LogOpts{
+		LoggerName:     programName,
+		FormatFunction: logFormat,
+	}
+
+	opts.Writers = []io.Writer{os.Stderr}
+	logger.Init(ctx, opts)
+
+	// TODO: prune old dirs
+
+	if err := refreshCreds(); err != nil {
+		logger.Errorf(err.Error())
+	}
+
+	logger.Infof("Done")
+}
+
+func refreshCreds() error {
+	project, err := getMetadata("project/project-id")
+	if err != nil {
+		return fmt.Errorf("Error getting project ID: %v", err)
+	}
+	domain := fmt.Sprintf("%s.svc.id.goog", project)
+	logger.Infof("Rotating workload credentials for domain %s", domain)
+
+	wisMd, err := getMetadata("instance/workload-identities")
+	if err != nil {
+		return fmt.Errorf("Error getting workload-identities: %v", err)
+	}
+
+	wtrcsMd, err := getMetadata("instance/workload-trusted-root-certs")
+	if err != nil {
+		return fmt.Errorf("Error getting workload-identities: %v", err)
+	}
+
+	wis := WorkloadIdentities{}
+	if err := json.Unmarshal(wisMd, &wis); err != nil {
+		return fmt.Errorf("Error unmarshaling workload trusted root certs: %v", err)
+	}
+
+	wtrcs := WorkloadTrustedRootCerts{}
+	if err := json.Unmarshal(wtrcsMd, &wtrcs); err != nil {
+		return fmt.Errorf("Error unmarshaling workload trusted root certs: %v", err)
+	}
+
+	now := time.Now().Format(time.RFC3339)
+	contentDir := fmt.Sprintf("%s-%s", contentDirPrefix, now)
+	tempSymlink := fmt.Sprintf("%s-%s", tempSymlinkPrefix, now)
+
+	logger.Infof("Creating timestamp contents dir %s", contentDir)
+
+	// TODO: validate filesystem permissions
+	if err := os.MkdirAll(contentDir, 0750); err != nil {
+		return fmt.Errorf("Error creating contents dir: %v", err)
+	}
+
+	if err := os.WriteFile(fmt.Sprintf("%s/certificates.pem", contentDir), []byte(wis.WorkloadCredentials[domain].CertificatePem), 0666); err != nil {
```

Comment:
why not `path.Join()` instead of `fmt.Sprintf()` ?

Reply:
`fmt` was already imported and we don't need multi-OS or input validation support which would warrant a path-specific function being used.

Reply:
It's a good practice to use `path.Join()` even if it's not a multi-OS solution - use the right api for the right job...

Reply:
i don't think that is a generally accepted belief. using '/' in fmt.Sprintf is extremely common in go for linux-specific code. the reasons to need a more specialized tool like the filepath package is when you need multi-OS, you need extraneous bits to be removed, you need filesystem-specific logic like relative path handling, etc.

Reply:
ok